plugins/inputs/dmcache/dmcache_linux.go

```diff
@@ -0,0 +1,180 @@
+// +build linux
+
+package dmcache
+
+import (
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"errors"
+
+	"github.com/influxdata/telegraf"
+)
+
+const metricName = "dmcache"
+
+type cacheStatus struct {
+	device            string
+	length            int
+	target            string
+	metadataBlocksize int
+	metadataUsed      int
+	metadataTotal     int
+	cacheBlocksize    int
+	cacheUsed         int
+	cacheTotal        int
+	readHits          int
+	readMisses        int
+	writeHits         int
+	writeMisses       int
+	demotions         int
+	promotions        int
+	dirty             int
+}
+
+func (c *DMCache) Gather(acc telegraf.Accumulator) error {
+	outputLines, err := c.getCurrentStatus()
+	if err != nil {
+		return err
+	}
+
+	total := make(map[string]interface{})
+
+	for _, s := range outputLines {
+		status, err := parseDMSetupStatus(s)
+		if err != nil {
+			return err
+		}
+
+		fields := make(map[string]interface{})
+		fields["length"] = status.length
+		fields["metadata_blocksize"] = status.metadataBlocksize
+		fields["metadata_used"] = status.metadataUsed
+		fields["metadata_total"] = status.metadataTotal
+		fields["cache_blocksize"] = status.cacheBlocksize
+		fields["cache_used"] = status.cacheUsed
+		fields["cache_total"] = status.cacheTotal
+		fields["read_hits"] = status.readHits
+		fields["read_misses"] = status.readMisses
+		fields["write_hits"] = status.writeHits
+		fields["write_misses"] = status.writeMisses
+		fields["demotions"] = status.demotions
+		fields["promotions"] = status.promotions
+		fields["dirty"] = status.dirty
+
+		if c.PerDevice {
+			tags := map[string]string{"device": status.device}
+			acc.AddFields(metricName, fields, tags)
+		}
+		aggregateStats(total, fields)
```

Comment:
Switch to fields at the last moment.  Overall flow should be get the cacheStatus, aggregate, then convert to fields and emit as needed.